File: cpp/Longest-Common-Prefix/Longest-Common-Prefix.cpp
  
/// Source : https://leetcode-cn.com/problems/longest-common-prefix/
/// Author : bryce
/// Time   : 2019-11-05
class Solution {
public:
    string longestCommonPrefix(vector<string>& strs) {
        int n = strs.size();
        int i = 0,b = 0;
        if (n == 0) return "";
        string res = "";
        int a = strs[0].size();
        for (i = 0; i < n; i++)
        {
            int m = strs[i].size();
            a = min(a,m);
            if (a >= m) b = i;
        }
        if (a==0) return "";
        res = strs[b];
        for (int l = 0; l < n; l++)
        {
            for(int j = 0; j < a;j++)
            {
                if (res[j] != strs[l][j])
                res = res.substr(0,j);           
            }
        }
        return res;
    }
};

File: python/Longest-Common-Prefix/Longest-Common-Prefix.py
#! /usr/bin/env python
#! -*- coding=utf-8 -*-
# Date: 2019-11-06
# Author: Bryce

# Python Solution 1:
class Solution:
    def longestCommonPrefix(self, strs):
        """
        :type strs: List[str]
        :rtype: str
        """
        res = ""
        for tmp in zip(*strs):              #zip ()打包 加星号是解压操作
            tmp_set = set(tmp)              #set() 函数创建一个无序不重复元素集，可进行关系测试，删除重复数据，还可以计算交集、差集、并集等。
            if len(tmp_set) == 1:           
                res += tmp[0]
            else:
                break
        return res
      
    # Python Solution 2:
    class Solution:
    def longestCommonPrefix(self, strs):
        if not strs:
            return ""
        res = strs[0]
        i = 1
        while i < len(strs):
            while strs[i].find(res) != 0:
                res = res[0:len(res)-1]
            i += 1
        return res
